### PR TITLE
Introduce SortMergeReader

### PR DESCRIPTION
**Pull request by LadyForest**
## What is the purpose of the change

This pull request introduces `SortMergeReader` and `SortMergeIterator` to perform a merge sort on SST files. The inputs and output are both ordered by key and sequence number. The merge process uses a min-heap of size k (k is the number of input readers) to maintain the iterator which points to the latest minimum key.

### CODE REVIEW

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            List<RecordReader<KeyValue>> sortedReaders, Comparator<RowData> userKeyComparator) {
+        this.sortedReaders = sortedReaders;
+        this.userKeyComparator = userKeyComparator;
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        List<RecordIterator<KeyValue>> sortedIterators = new ArrayList<>();
+        for (RecordReader<KeyValue> reader : sortedReaders) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: Consider merging two reader:
- reader1: batch(2, 3) batch(4)
- reader2: batch(1, 5)

The correct result is: 1, 2, 3, 4, 5
But looks like the result should be: 1, 2, 3, 5, 4
What do you think?

Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: You're absolutely right. Sorting should consider batch crossing.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import java.util.Objects;
+import java.util.PriorityQueue;
+
+/**
+ * This reader is to read a list of {@link RecordReader}, which is already sorted by key and
+ * sequence number, and perform a sort merge algorithm. Note that values are not combined at this
+ * phase.
+ */
+public class SortMergeReader implements RecordReader<KeyValue> {
+    private final List<RecordReader<KeyValue>> sortedReaders;
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java:40]: Create a empty line up.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    private final Comparator<RowData> userKeyComparator;
+    private PriorityQueue<Element> minHeap;
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java:43]: left a empty line between final and non-final

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        this.userKeyComparator = userKeyComparator;
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        if (sortedReaders.isEmpty()) {
+            return null;
+        }
+        if (Objects.isNull(minHeap)) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: minHeap == null

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            heapify();
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: initMinHeap?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                            if (result != 0) {
+                                return result;
+                            }
+                            return Long.compare(
+                                    e1.iterator.current().sequenceNumber(),
+                                    e2.iterator.current().sequenceNumber());
+                        });
+        for (int i = 0; i < sortedReaders.size(); i++) {
+            RecordIterator<KeyValue> iterator = sortedReaders.get(i).readBatch();
+            if (Objects.nonNull(iterator) && iterator.advanceNext()) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: Objects.nonNull and isNull is used used as a {@link java.util.function.Predicate}, {@code filter(Objects::isNull)}. (In method document)

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    /** The iterator iterates on {@link SortMergeReader}. */
+    private class SortMergeIterator implements RecordIterator<KeyValue> {
+        private KeyValue previous;
+
+        @Override
+        public boolean advanceNext() throws IOException {
+            if (minHeap.size() > 0) {
+                current = minHeap.poll();
+                previous = current.iterator.current();
+                if (current.iterator.advanceNext()) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: I feel that there is a problem here, after `advanceNext`, it is not necessarily the current minimum ah.
(Note the test coverage)

**Review by JingsongLi — CHANGES_REQUESTED** [review]

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java (hunk 5 above)
Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: +1. 
Take `PriorityQueue#initElementsFromCollection` for reference, change `heapify` to `initElements` and `supply` to `supplyElements` maybe better.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java (hunk 7 above)
Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: Yes, according to the current impl, only previous is ordered. I'll change it.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+    @Nullable
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        if (sortedReaders.isEmpty()) {
+            return null;
+        }
+        if (minHeap == null) {
+            initElements();
+        } else if (minHeap.size() > 0) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: I don't fully understand this logic. `minHeap.size() > 0`?
Why not just invoke `topElement.readBatch`?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        RecordIterator<KeyValue> iterator = reader.readBatch();
+        if (iterator != null && iterator.advanceNext()) {
+            minHeap.offer(new Element(iterator, topElement.index));
+        }
+    }
+
+    /** The iterator iterates on {@link SortMergeReader}. */
+    private class SortMergeIterator implements RecordIterator<KeyValue> {
+
+        private boolean innerIteratorsAdvanceAhead = true;
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: just `readPeek`?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+                    minHeap.offer(new Element(topIterator, topElement.index));
+                    topElement = minHeap.peek();
+                    return true;
+                } else {
+                    // reach the end of current batch, should invoke the next batch
+                    return false;
+                }
+            } else if (topElement != null) {
+                // keep pace with the rest long tail iterator
+                return topElement.iterator.advanceNext();
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: Why here? should not end of reader?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+ * sequence number, and perform a sort merge algorithm. Note that values are not combined at this
+ * phase.
+ */
+public class SortMergeReader implements RecordReader<KeyValue> {
+
+    private final List<RecordReader<KeyValue>> sortedReaders;
+    private final Comparator<RowData> userKeyComparator;
+
+    private PriorityQueue<Element> minHeap;
+    private Element topElement;
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: `minElement`?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java (hunk 8 above)
Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: Correct me if I'm wrong. If there's just one reader left (a long-tail reader), I think there's no need to offer the element to heap. So the condition for supplying elements is that the heap is not empty.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java (hunk 10 above)
Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: If we offer all elements to the heap, there's no need to do this. (For better code readability) However, if there's one long-tail reader left and we don't choose to offer the element to the heap, this check is needed I think.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java (hunk 9 above)
Comment (LadyForest) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: maybe `justPeek` or `peekOnly`?

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java [hunk 12]
```diff
@@ -133,4 +146,154 @@ public KeyValue current() {
...
+                return true;
+            }
+
+            if (!(o instanceof TestKeyValue)) {
+                return false;
+            }
+
+            TestKeyValue that = (TestKeyValue) o;
+
+            return new EqualsBuilder()
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java]: I think we can just use equals generated by intellij:
```
return sequenceNumber == that. sequenceNumber
                && key == that. key
                .......;
```

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java [hunk 13]
```diff
@@ -133,4 +146,154 @@ public KeyValue current() {
...
+                    .append(sequenceNumber, that.sequenceNumber)
+                    .append(key, that.key)
+                    .append(valueKind, that.valueKind)
+                    .append(value, that.value)
+                    .isEquals();
+        }
+
+        @Override
+        public int hashCode() {
+            return new HashCodeBuilder(17, 37)
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java]: `Objects.hash(...)` is better.

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java [hunk 14]
```diff
@@ -133,4 +146,154 @@ public KeyValue current() {
...
+            throws IOException {
+        List<RecordReader<KeyValue>> readers = new ArrayList<>();
+        for (List<TestKeyValue> input : inputs) {
+            readers.add(new TestRecordReader(input, batchSize));
+        }
+        RecordReader<KeyValue> sortMergeReader = new SortMergeReader(readers, USER_KEY_COMPARATOR);
+        List<TestKeyValue> values = new ArrayList<>();
+        RecordIterator<KeyValue> batch;
+        while ((batch = sortMergeReader.readBatch()) != null) {
+            while (batch.advanceNext()) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java:276]: Do you have any case to cover object reuse?
For example, you can add test here:
```
KeyValue previous = null;
while (batch.advanceNext()) {
     if (previous != null) {
          Assert.assertEqual(previous, values.getLastOne);
     }
     previous = batch.current();
     ......
}
```

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java [hunk 15]
```diff
@@ -133,4 +146,154 @@ public KeyValue current() {
...
+                values.add(TestKeyValue.from(combinePolicy, batch));
+            }
+        }
+        combineReader.close();
+        Assert.assertTrue(reader.isClosed());
+        Assert.assertEquals(expected, values);
+    }
+
+    public static void testSortMerge(List<List<TestKeyValue>> inputs, int batchSize)
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java:264]: Can you add a `testSortMergeWithCombiner`:
`List<List<TestKeyValue>> inputs, int batchSize, CombinePolicy combinePolicy, List<TestKeyValue> expected`,

**Review by JingsongLi — COMMENTED** [review]
Final left some comments about testing.

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java (hunk 12 above)
Comment (LadyForest) [reply flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/utils/TestRecordReader.java]: Sure. Btw this code snippet is also auto-generated by IntelliJ, out of the different strategy...

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+                    // keep pace with the inner iterators
+                    justPeek = false;
+                    minElement = minHeap.peek();
+                    return true;
+                }
+                minElement = minHeap.poll();
+                // find the top iterator to advance
+                RecordIterator<KeyValue> topIterator = minElement.iterator;
+                if (topIterator.advanceNext()) {
+                    minHeap.offer(new Element(topIterator, minElement.index));
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java]: Why not just `minElement`?

**Review by JingsongLi — APPROVED** [review]
Thanks for the update, looks good to me!

**Pull request merged**